Sof 1640 internationalize sofie

temk-tv2 opened the pull request:
Introduce Angular Internationalization.
Support en and da translations.
Add locale switcher, which is not going to work on development mode. However we need to setup the deployment configuration and test it in production.

LindvedKrvang commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,187 @@
+      </trans-unit>
+      <trans-unit id="2175030815390487298" datatype="html">
+        <source>rundown-header-context-menu-component.deactivate.button</source>
+        <target>Deaktivere</target>
```
`Deaktiver`

LindvedKrvang commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,187 @@
+      </trans-unit>
+      <trans-unit id="1013806782562403538" datatype="html">
+        <source>rundown-header-context-menu-component.take.button</source>
+        <target>Take</target>
```
Are we also calling it a "Take" in danish?

andr9528 replied:
Perhabs it should be "Tag"? 
Sounds weird, but should be the direct translation.

LindvedKrvang replied:
I think we should ask Jacob what he prefers

temk-tv2 replied:
I think the best is to leave it as "take"

LindvedKrvang commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,187 @@
+      </trans-unit>
+      <trans-unit id="2159130950882492111" datatype="html">
+        <source>Cancel</source>
+        <target>Annullere</target>
```
`Anuller`

KvelaGorrrrnio replied:
`Annuller` or `Fortryd`.

LindvedKrvang replied:
`Annuller`. `Fortryd` indicates you already did an action you want to undo i.e. do another action to remedy the first action.

LindvedKrvang commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,187 @@
+      </trans-unit>
+      <trans-unit id="1079134994533846284" datatype="html">
+        <source>Lost connection to backend. Attempting to reconnect...</source>
+        <target>Mistet forbindelse til backend. Forsøger at genoprette forbindelse...</target>
```
I think that instead of `backend` we should call it `server`. Most non-tech people don't know what a `backend` is.

LindvedKrvang reviewed (commented):
Looks good.
Only a few questions.

I guess you can mostly ignore the comments on the danish translations since we are not planning on using danish that much?

temk-tv2 commented on src/app/shared/components/confirmation-dialog/confirmation-dialog.component.ts:
```diff
@@ -24,10 +24,10 @@ export interface ConfirmationDialogData {
-  public title: string = 'Are you sure?'
+  public title: string = $localize`:confirmation-dialog.are-you-sure.title:Are you sure?`
   public message: string = ''
-  public okButtonText: string = 'Yes'
-  public cancelButtonText: string = 'Cancel'
+  public okButtonText: string = $localize`:confirmation-dialog.yes.button:Yes`
+  public cancelButtonText: string = $localize`:confirmation-dialog.cancel.button:Cancel`
```
yes, those are default values.

KvelaGorrrrnio replied:
It is not possible to have it fallback to the english translation file?

LindvedKrvang replied:
I would hope so

temk-tv2 commented on src/app/rundown-overview/components/rundown-overview/rundown-overview.component.ts:
```diff
@@ -46,8 +46,11 @@ export class RundownOverviewComponent implements OnInit, OnDestroy {
-      this.deleteRundown(basicRundown.id)
+    this.dialogService.createConfirmDialog(
+      $localize`rundown-overview-component.deletion-dialog-tittle:Delete Rundown?`,
+      $localize`rundown-overview-component.deletion-dialog-question:Are you sure you want to delete the Rundown "${basicRundown.name}"?\n\nPlease note: This action is irreversible!`,
```
I probably forgot to delete the actual text. It doesn't seem to be the merge.

KvelaGorrrrnio commented on src/app/shared/components/locale-switcher/locale-switcher.component.ts:
```diff
@@ -0,0 +1,19 @@
+  constructor(@Inject(LOCALE_ID) public activeLocale: string) {}
+
+  public onChange(): void {
+    window.location.href = `/${this.activeLocale}`
```
Shouldn't it redirect to the current page in the new locale?

temk-tv2 replied:
After some discussions with you and Lindved, I have decided to remove the locale-switcher feature, since we anyway wont be able to use it for now.

KvelaGorrrrnio commented on src/app/rundown-overview/components/rundown-overview/rundown-overview.component.ts:
```diff
@@ -46,8 +46,11 @@ export class RundownOverviewComponent implements OnInit, OnDestroy {
-    this.dialogService.createConfirmDialog('Delete Rundown?', `Are you sure you want to delete the Rundown "${basicRundown.name}"?\n\nPlease note: This action is irreversible!`, 'Delete', () =>
-      this.deleteRundown(basicRundown.id)
+    this.dialogService.createConfirmDialog(
+      $localize`rundown-overview-component.deletion-dialog-tittle`,
```
`tittle` is misspelled.

andr9528 commented on src/app/rundown-overview/components/rundown-overview/rundown-overview.component.html:
```diff
@@ -1,15 +1,15 @@
 <sofie-header></sofie-header>
 
-<h1 class="c-rundown-overview__title">Rundowns</h1>
+<h1 class="c-rundown-overview__title" i18n>rundown-overview-component.rundown-header.label</h1>
```
With that example, how would a key used in a service be named?
Doesn't appear to be many instances were that is needed at the moment, but could become relevant. 
At the moment i onlly really see one in DialogService.

KvelaGorrrrnio replied:
@andr9528 My suggestion would be `<module-name>.<class-name>.<local-key>`. So for the dialog-service:
`shared.dialog-service.my-translation-key`.

temk-tv2 replied:
to keep the translation key shorter and precise, my suggestion would be `<component-name>.<attribute>.<local-key>`
as it is in here:
`rundown-overview-component.rundown-header.label`

KvelaGorrrrnio replied:
`rundown-header` is misleading. The label is placed at the top of the page, but it is not in the header. The sofie-header is a component that is used right above this translation. And we have a component called rundown header in the rundown module. So there might be some confusion when a translator only looks at the translation entries without the corresponding code.

By having the class name as the top-level domain, we might have conflicts if we need to do translations in classes/components with the same name across modules. For now it should not be a problem, but as the project grows and more pages and modules are added, the likelihood increases accordingly.

Both suggestions use 2 levels of categorization, so they should be equivalently short and precise.

KvelaGorrrrnio replied:
What does the thumbs up indicate in this case? 😄

KvelaGorrrrnio commented on angular.json:
```diff
@@ -76,19 +96,27 @@
           "builder": "@angular-devkit/build-angular:dev-server",
           "configurations": {
+            "da": {
+              "browserTarget": "ng-sofie:build:development,da"
+            },
+            "en": {
+              "browserTarget": "ng-sofie:build:development,en"
+            },
             "production": {
               "browserTarget": "ng-sofie:build:production"
             },
             "development": {
               "browserTarget": "ng-sofie:build:development"
             }
           },
-          "defaultConfiguration": "development"
+          "defaultConfiguration": "en"
```
Do we need to have both danish and english for development purposes?
An alternative could be to keep the default configuration as development and then update production and development to have the browserTarget of the `en` configuration.

temk-tv2 replied:
I have made the suggested changes.

KvelaGorrrrnio commented on package.json:
```diff
@@ -3,7 +3,7 @@
   "scripts": {
     "ng": "ng",
-    "start": "ng serve",
+    "start": "ng serve --configuration=en",
```
The configuration flag can be removed, since the default configuration is set in angular.json.

temk-tv2 replied:
I have made the suggested changes.

RasmusAlbrektsen commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,195 @@
+      </trans-unit>
+      <trans-unit id="672684067026346719" datatype="html">
+        <source>dialog.cancel-button:Cancel</source>
+        <target>Annulere</target>
```
Skal være `Annuller`

RasmusAlbrektsen commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,195 @@
+      </trans-unit>
+      <trans-unit id="3796920100771349539" datatype="html">
+        <source>rundown-overview-component.deletion-dialog-question &quot;<x id="PH" equiv-text="basicRundown.name"/>&quot;? Please note: This action is irreversible!</source>
+        <target>Er du sikker at du vil slette Rundown &quot;<x id="PH" equiv-text="basicRundown.name"/>&quot;?Vær opmærksom på: Denne handling er uigenkaldelig!</target>
```
Der mangler et "på" i denne sætning:
`Er du sikker på at du vil slette Rundown`

LindvedKrvang commented on src/app/rundown-overview/components/rundown-overview/rundown-overview.component.ts:
```diff
@@ -46,8 +46,11 @@ export class RundownOverviewComponent implements OnInit, OnDestroy {
-      this.deleteRundown(basicRundown.id)
+    this.dialogService.createConfirmDialog(
+      $localize`rundown-overview-component.deletion-dialog-tittle:Delete Rundown?`,
+      $localize`rundown-overview-component.deletion-dialog-question:Are you sure you want to delete the Rundown "${basicRundown.name}"?\n\nPlease note: This action is irreversible!`,
```
I'm in favour for the latter

temk-tv2 replied:
I am afraid that the syntax won't be able to be applied since I am getting an error saying that TypeScript is expecting a TemplateStringsArray type, and not a plain string.

temk-tv2 commented on src/locale/messages.da.xlf:
As i18n translations are quite new to me, I have googled the question and here is a resume of good practices:

- **Adding a Translation Key**: When adding a new translation key, make sure it is added to all relevant translation files for supported languages. You can create the key in your source language (usually the default language of your application) and then provide translations in other languages.

**Removing a Translation Key**: If a translation key is no longer needed, it's a good practice to mark it as deprecated instead of removing it immediately. This allows you to maintain backward compatibility for older versions of your application that might still rely on the key. After a reasonable deprecation period, you can safely remove it.
**Changing a Translation Key**: Avoid changing existing translation keys if possible, as this can lead to inconsistencies and issues with older versions of your application. If you need to make changes, consider creating a new key and marking the old one as deprecated.

KvelaGorrrrnio reviewed (commented):
Since we still have an open thread for the translation key structure, I haven't looked at the translation keys yet.

temk-tv2 commented:
@KvelaGorrrrnio I have addressed your comments, also implemented the merge feature so we dont have to delete/ deprecate translations in case we change them.

RasmusAlbrektsen commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,67 @@
+      </trans-unit>
+      <trans-unit id="672684067026346719" datatype="html">
+        <source>dialog.cancel-button:Cancel</source>
+        <target>Annuler</target>
```
`Annuller`

RasmusAlbrektsen commented on src/locale/messages.da.xlf:
```diff
@@ -0,0 +1,67 @@
+      </trans-unit>
+      <trans-unit id="3796920100771349539" datatype="html">
+        <source>rundown-overview-component.deletion-dialog-question &quot;<x id="PH" equiv-text="basicRundown.name"/>&quot;</source>
+        <target state="new">Er du sikker på at du vil slette Rundown &quot;<x id="PH" equiv-text="basicRundown.name"/>&quot;?Vær opmærksom på: Denne handling er uigenkaldelig!</target>
```
Rather than `uigenkaldelig` i would use `uoprettelig` since it refers to how it is not possible to rectify what is caused by the operation.

KvelaGorrrrnio commented on angular.json:
```diff
@@ -4,6 +4,17 @@
+          "da": {
+            "translation": "src/locale/messages.da.xlf"
+          },
+          "en": {
```
This hasn't been updated yet :)

KvelaGorrrrnio reviewed (commented):
A minor thing. When writing commit messages, it is unnecessary to write that the changes made are from PR comments – just write what the change is 😃

KvelaGorrrrnio reviewed (commented):
I have added a minor fix for the angular.json configuration to get yarn start to work.
But when I start the application I only see translation keys as depicted below.

![Skærmbillede 2023-11-17 kl  14 25 11](https://github.com/tv2/ng-sofie/assets/11258272/c4b3521b-4898-4962-947d-5e47ea60cf97)

temk-tv2 commented:
@KvelaGorrrrnio  I have now fixed the problem with the source locale.
